Develop1

DreamBreezz opened the pull request:
Покрыл тестами 4 класса проекта, согласно заданию дипломного проекта.

DreamBreezz committed:
update: pom.xml
update: README.MD
add: BunTests.java
add: IngredientTests.java
add: IngredientTypeTests.java

DreamBreezz committed:
add: Constants.java
update: BunTests.java
update: IngredientTypeTests.java

DreamBreezz committed:
update: BunTests.java
update: IngredientTests.java (add parameterization)

DreamBreezz committed:
add: BurgerTests.java (but need fix!)

DreamBreezz committed:
fix: BurgerTests.java (StringBuilder added)

DreamBreezz committed:
refactor: BurgerTests.java (StringBuilder removed)

DreamBreezz committed:
refactor: BurgerTests.java (add createMockedBurger method)

DreamBreezz committed:
minor changes
add: jacoco report

pomegranate282922 commented on src/test/java/praktikum/BunTests.java:
```diff
@@ -0,0 +1,25 @@
+
+    @Test
+    public void bunGetPriceTest() {
+        assertEquals(bun.getPrice(), bunPrice, 0.001);
```
Дельту лучше вынести как константу 
Можно улучшить

DreamBreezz replied:
done

pomegranate282922 commented on src/test/java/praktikum/BurgerTests.java:
```diff
@@ -0,0 +1,107 @@
+    @Test
+    public void burgerSetBunsTest() {
+        burger.setBuns(bunMock);
+        Mockito.verify(burger).setBuns(bunMock);
```
Здесь проверяется что если метод вызвать, он вызвался 
Зачем?
Нужно исправить

DreamBreezz replied:
done, убрал бессмысленную проверку

pomegranate282922 commented on src/test/java/praktikum/IngredientTests.java:
```diff
@@ -0,0 +1,36 @@
+    @Test
+    public void ingredientIngredientTypeTest() {
+        Ingredient i = new Ingredient(type, name, price);
+        System.out.println(i.getType() + ": " + i.getName() + ", " + i.getPrice()); //чтобы хоть что-то в консоли видеть
```
Вывод в консоль лучше убрать 
Можно улучшить

DreamBreezz replied:
тут не согласен. Но убрал))

pomegranate282922 commented on src/test/java/praktikum/IngredientTypeTests.java:
```diff
@@ -0,0 +1,18 @@
+    @Test
+    public void isFillingFillingTest() {
+        assertEquals(IngredientType.FILLING.toString(), "FILLING");
+    }
```
Здесь тоже просится параметризация 
Можно улучшить

DreamBreezz replied:
честно говоря, вообще не понимаю, зачем тестировать enum-класс. Но сделал и с параметризацией))

pomegranate282922 reviewed (commented):
Работа хорошая! Все тесты реализованы. Есть моки и параметризация. 
Замечания: в бургер тестах убрать проверки ни о чем. Тестируем результаты метода.

DreamBreezz committed:
upd: fixes from review

DreamBreezz committed:
upd: fixes from review

pomegranate282922 commented on src/test/java/praktikum/BurgerTests.java:
```diff
@@ -0,0 +1,101 @@
+        burger.addIngredient(ingredientMock);
+        burger.moveIngredient(0, 1);
+        burger.removeIngredient(0);
+        Mockito.verify(burger, times(2)).addIngredient(ingredientMock);
```
Так. Здесь снова проверяем работу библиотеки мокито) или что если вызывать метод, он вызовется..
Нужно проверить реальное поведение всех методов класса бургер , используя моки для зависимости класса. 
нужно исправить

DreamBreezz replied:
Сделал вместо этого два теста, под каждый метод (move, remove) свой тест, с моками

pomegranate282922 reviewed (commented):
Так, осталось важное замечание - его нужно осознать и исправить.

DreamBreezz committed:
upd: BurgerTests

DreamBreezz committed:
upd: Jacoco report

DreamBreezz committed:
upd: Jacoco report

pomegranate282922 commented on src/test/java/praktikum/BurgerTests.java:
```diff
@@ -0,0 +1,109 @@
+        burger.addIngredient(ingredientMock);
+        burger.addIngredient(ingredientMock2);
+        burger.moveIngredient(2,0);
+        Assert.assertEquals(burger.ingredients.get(0), ingredientMock2);
```
Сначала ожидаемое значение, потом актуальное 
Нужно исправить

DreamBreezz replied:
done

pomegranate282922 commented on src/test/java/praktikum/BurgerTests.java:
```diff
@@ -0,0 +1,109 @@
+        Mockito.verify(bunMock, times(2)).getName();
+        Mockito.verify(ingredientMock, times(1)).getType(); // цифра = кол-во вызовов
+        Mockito.verify(ingredientMock, times(1)).getName(); //         addIngredient
+        Mockito.verify(burger, times(1)).getPrice();
```
Это тоже нужно убрать

DreamBreezz replied:
done, убрал тест

pomegranate282922 commented on src/test/java/praktikum/BurgerTests.java:
```diff
@@ -0,0 +1,109 @@
+    Ingredient ingredientMock2;
+
+    @Test
+    public void burgerMoveIngredientTest() {
```
У бургера не все методы протестированы. Нужно дополнить,

DreamBreezz replied:
done, добавил burgerSetBunsTest и burgerAddIngredientTest.

Но вообще, не согласен, эти два теста получаются какой-то оверхед, ибо в первом мы тестируем обычный сеттер, а во втором вообще скорее класс List, чем Burger
И без отдельных тестов на эти методы отчёт показывал 100% покрытие:
<img width="560" alt="image" src="https://github.com/user-attachments/assets/a5d3dad6-5dc8-44a0-b26d-26bd39a1feee">
т.к. энивей setBuns и addIngredient вызываются в других тестах этого класса.
К тому же, в формальном задании к первой части диплома и не требуется обязательно покрывать все методы, минимальный процент покрытия 70%.

pomegranate282922 replied:
Ок)
Главное - разобраться!

pomegranate282922 commented on src/test/java/praktikum/BurgerTests.java:
```diff
@@ -0,0 +1,109 @@
+@RunWith(MockitoJUnitRunner.class)
+public class BurgerTests {
+
+    @Spy
```
Зачем шпион?
Нужно исправить

DreamBreezz replied:
done, заменил на обычный объект

но вообще, тоже не согласен – как минимум, я бы хотел быть уверен, что в методе Burger.getPrice вызываются Bun.getPrice и Ingredient.getPrice

pomegranate282922 replied:
Это хорошо проверить и правильно! Но внимание вопрос - зачем для этого шпион у бургера?)

pomegranate282922 reviewed (commented):
И все-таки не до конца доделано, хотя суть уже правильная, конечно. Прошу поправить важные замечания.

DreamBreezz committed:
fix from review, upd: BurgerTests:
- delete getReceiptVerifyTest,
- delete @Spy burger,
- update burgerMoveIngredientTest,
- update burgerGetPriceTest,
- add burgerAddIngredientTest,
- add burgerSetBunsTest

upd: Jacoco report

pomegranate282922 reviewed (approved):
Отлично, зачет! 
а со шпионом еще лучше почитать.
